### PR TITLE
Address/LegacyAddress/SegwitAddress: construct with and store Network (not NetworkParameters)

### PR DESCRIPTION
**Pull request by msgilligan**
* Address/LegacyAddress/SegwitAddress: construct with and store Network (not NetworkParameters)
* Replace static factories that take NetworkParameters
* Add static factory methods that take `Network`
* Deprecate static factory methods that take `NetworkParameters`
* Update all (non-test) usages to use the new methods

Note that we had to disable an alt-net test in LegacyAddressTest (we should bet able re-enable once PR #2608 is merged and we rebase on top of it.)

### CODE REVIEW

**Comment by schildbach** [comment]
> Note that we had to disable an alt-net test in LegacyAddressTest (we should bet able re-enable once PR #2608 is merged and we rebase on top of it.)

#2608 is merged.

**Comment by msgilligan** [comment]
> > Note that we had to disable an alt-net test in LegacyAddressTest (we should bet able re-enable once PR #2608 is merged and we rebase on top of it.)
> 
> #2608 is merged.

I tried re-enabling the disabled test and it fails with `java.lang.ClassCastException` trying to cast `BitcoinNetwork` to `Network`. PR #2501 will fix the issue, so I'll finish the rebase/force-push once PR #2501 is merged.

File: core/src/main/java/org/bitcoinj/core/NetworkParameters.java [hunk 1]
```diff
@@ -176,17 +177,17 @@ public static NetworkParameters fromID(String id) {
...
      */
     public static NetworkParameters of(Network network) {
-        if (network instanceof BitcoinNetwork) {
-            return BitcoinNetworkParams.of((BitcoinNetwork) network);
-        } else {
-                throw new IllegalArgumentException("Unknown network");
-        }
+        return (network instanceof BitcoinNetwork)
+                ? BitcoinNetworkParams.of((BitcoinNetwork) network)
+                : Networks.find(network).orElseThrow(() -> new IllegalArgumentException("Unknown network"));
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/NetworkParameters.java:190]: Is this change strictly needed in this commit, or could it be separated?

File: core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java [hunk 2]
```diff
@@ -437,13 +437,13 @@ public Address freshAddress(KeyChain.KeyPurpose purpose) {
...
             Script outputScript = chain.freshOutputScript(purpose);
             checkState(ScriptPattern.isP2SH(outputScript)); // Only handle P2SH for now
-            Address freshAddress = LegacyAddress.fromScriptHash(params,
+            Address freshAddress = LegacyAddress.fromScriptHash(params.network(),
                     ScriptPattern.extractHashFromP2SH(outputScript));
             maybeLookaheadScripts();
             currentAddresses.put(purpose, freshAddress);
             return freshAddress;
         } else if (outputScriptType == ScriptType.P2PKH || outputScriptType == ScriptType.P2WPKH) {
-            return Address.fromKey(params, freshKey(purpose), outputScriptType);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java:446]: There is a deprecated API usage migrationl snuck in here. I'll prepare a commit that migrates all of them.

File: core/src/test/java/org/bitcoinj/uri/BitcoinURITest.java [hunk 3]
```diff
@@ -85,8 +86,13 @@ public void testConvertToBitcoinURI() {
...
 
-        assertEquals("mockcoin:" + mockNetGoodAddress + "?amount=12.34&label=Hello&message=AMessage",
-             BitcoinURI.convertToBitcoinURI(LegacyAddress.fromBase58(alternativeParameters, mockNetGoodAddress), parseCoin("12.34"), "Hello", "AMessage"));
+        Networks.register(alternativeParameters);
+        try {
+            assertEquals("mockcoin:" + mockNetGoodAddress + "?amount=12.34&label=Hello&message=AMessage",
+                    BitcoinURI.convertToBitcoinURI(LegacyAddress.fromBase58(alternativeParameters.network(), mockNetGoodAddress), parseCoin("12.34"), "Hello", "AMessage"));
+        } finally {
+            Networks.unregister(alternativeParameters);
+        }
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/uri/BitcoinURITest.java:95]: I sense a deja-vu here (-:

Probably also a candidate for separation.

**Review by schildbach — COMMENTED** [review]
Generally looks good. Review comments inline.

File: core/src/main/java/org/bitcoinj/script/Script.java [hunk 4]
```diff
@@ -393,16 +395,17 @@ public Address getToAddress(NetworkParameters params) throws ScriptException {
...
+        Network network = params.network();
         if (ScriptPattern.isP2PKH(this))
-            return LegacyAddress.fromPubKeyHash(params, ScriptPattern.extractHashFromP2PKH(this));
+            return LegacyAddress.fromPubKeyHash(network, ScriptPattern.extractHashFromP2PKH(this));
         else if (ScriptPattern.isP2SH(this))
-            return LegacyAddress.fromScriptHash(params, ScriptPattern.extractHashFromP2SH(this));
+            return LegacyAddress.fromScriptHash(network, ScriptPattern.extractHashFromP2SH(this));
         else if (forcePayToPubKey && ScriptPattern.isP2PK(this))
-            return LegacyAddress.fromKey(params, ECKey.fromPublicOnly(ScriptPattern.extractKeyFromP2PK(this)));
+            return LegacyAddress.fromPubKeyHash(network, ECKey.fromPublicOnly(ScriptPattern.extractKeyFromP2PK(this)).getPubKeyHash());
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/script/Script.java:404]: Just noticed this change, is it correct?

File: core/src/main/java/org/bitcoinj/core/NetworkParameters.java (hunk 1 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/core/NetworkParameters.java:190]: It is required for this commit to work properly. (If an address is created from an alt-`Network`, we need to be able to retrieve the corresponding alt-`NetworkParameters`.)  However, it could be placed in a separate commit.

File: core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java (hunk 2 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java:446]: OK, thanks!

File: core/src/test/java/org/bitcoinj/uri/BitcoinURITest.java (hunk 3 above)
Comment (msgilligan) [reply core/src/test/java/org/bitcoinj/uri/BitcoinURITest.java:95]: Yeah, this can be done separately, although the register/unregister is not necessary when `LegacyAddress` is storing `NetworkParameters`.

File: core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java (hunk 2 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/wallet/KeyChainGroup.java:446]: See https://github.com/bitcoinj/bitcoinj/pull/2616

File: core/src/main/java/org/bitcoinj/script/Script.java (hunk 4 above)
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/script/Script.java:404]: Yes, the functionality is equivalent. Since `fromKey` was deprecated, `fromPubKeyHash` achieves the same result. I think it would also be possible to use `ECKey.toAddress()` but using a `static` method of `LegacyAddress` matched the other cases more closely.

Comment (schildbach) [reply core/src/main/java/org/bitcoinj/script/Script.java:404]: To clarify, this switches from `fromKey` to `fromPubKeyHash`.

Comment (schildbach) [reply core/src/main/java/org/bitcoinj/script/Script.java:404]: Ok thanks!

**Comment by schildbach** [comment]
I updated all the usages in tests too, and merged!
